# Handwritten Digit Recognition using CNN (MNIST)

This project implements a Convolutional Neural Network (CNN) to recognise handwritten digits (0–9) using the MNIST dataset.

The model is trained using Keras and TensorFlow and achieves high accuracy on test data. A GUI is also built using Tkinter to allow users to draw digits and get predictions in real-time.

In [1]:
pip install numpy tensorflow keras pillow

Note: you may need to restart the kernel to use updated packages.


***

## Importing Required Libraries

In this step, we import all the necessary libraries required for building and training the deep learning model.

- `keras` and `tensorflow` for building the CNN
- `numpy` for numerical operations
- `PIL` for image processing
- `tkinter` for building the GUI application

## Loading the MNIST Dataset

The MNIST dataset contains 60,000 training images and 10,000 testing images of handwritten digits (0–9).

Each image is grayscale and has a size of 28x28 pixels.

In [5]:
import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as K
# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()
print(x_train.shape, y_train.shape)

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 2s 0us/step
(60000, 28, 28) (60000,)


***

## Data Preprocessing

Before feeding the data into the model, we perform the following steps:

- Reshape the dataset to include a channel dimension (required for CNN)
- Normalize pixel values (scale from 0–255 to 0–1)
- Convert labels into categorical format using one-hot encoding

In [17]:
x_train = x_train.reshape(x_train.shape[0], 28, 28, 1)
x_test = x_test.reshape(x_test.shape[0], 28, 28, 1)
input_shape = (28, 28, 1)
# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples


***

## Building the CNN Model

We build a Convolutional Neural Network with the following layers:

- Convolutional layers to extract features
- MaxPooling layers to reduce dimensionality
- Dropout layers to prevent overfitting
- Fully connected (Dense) layers for classification

The final layer uses **Softmax activation** to classify digits into 10 categories (0–9).

In [15]:
batch_size = 128
num_classes = 10
epochs = 10
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),activation='relu',input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(256, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))
model.compile(loss=keras.losses.categorical_crossentropy,optimizer=keras.optimizers.Adadelta(),metrics=['accuracy'])

C:\Users\manan\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


***

## Training the Model

The model is trained on the training dataset using:

- Batch size: 128
- Epochs: 10

Validation is performed on the test dataset to monitor performance.

## Saving the Model

After training, the model is saved for future use.

This allows us to load the trained model later without retraining.

In [19]:
hist = model.fit(x_train, y_train,batch_size=batch_size,epochs=epochs,verbose=1,validation_data=(x_test, y_test))
print("The model has successfully trained")
model.save('mnist.h5')
print("Saving the model as mnist.h5")

Epoch 1/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 70s 145ms/step - accuracy: 0.1536 - loss: 2.2691 - val_accuracy: 0.4507 - val_loss: 2.2002
Epoch 2/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 67s 144ms/step - accuracy: 0.3291 - loss: 2.1579 - val_accuracy: 0.6118 - val_loss: 2.0608
Epoch 3/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 68s 146ms/step - accuracy: 0.4665 - loss: 2.0111 - val_accuracy: 0.6895 - val_loss: 1.8705
Epoch 4/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 69s 146ms/step - accuracy: 0.5580 - loss: 1.8162 - val_accuracy: 0.7456 - val_loss: 1.6260
Epoch 5/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 69s 147ms/step - accuracy: 0.6218 - loss: 1.5844 - val_accuracy: 0.7790 - val_loss: 1.3569
Epoch 6/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 68s 146ms/step - accuracy: 0.6627 - loss: 1.3568 - val_accuracy: 0.8005 - val_loss: 1.1120
Epoch 7/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 72s 154ms/step - accuracy: 0.6899 - loss: 1.1697 - val_accuracy: 0.8179 - val_loss: 0.9229
Epoch 8/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 71s 151ms/step - accuracy: 0.7164 - loss: 1

The model has successfully trained
Saving the model as mnist.h5


***

## Model Evaluation

The model is evaluated on the test dataset to measure its performance.

Metrics:
- Test Loss
- Test Accuracy

In [21]:
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test loss: 0.622086763381958
Test accuracy: 0.8522999882698059


***

## Building a GUI for Digit Recognition

A graphical user interface (GUI) is built using Tkinter.

Features:
- Draw digits using mouse
- Click "Recognise" to predict the digit
- Displays prediction with confidence score

## Prediction Function

The function processes the drawn image by:

- Resizing to 28x28 pixels
- Converting to grayscale
- Normalizing pixel values
- Passing it through the trained model

It returns:
- Predicted digit
- Confidence score

In [23]:
from keras.models import load_model
from tkinter import *
import tkinter as tk
import win32gui
from PIL import ImageGrab, Image
import numpy as np
model = load_model('mnist.h5')
def predict_digit(img):
    #resize image to 28x28 pixels
    img = img.resize((28,28))
    #convert rgb to grayscale
    img = img.convert('L')
    img = np.array(img)
    #reshaping to support our model input and normalizing
    img = img.reshape(1,28,28,1)
    img = img/255.0
    #predicting the class
    res = model.predict([img])[0]
    return np.argmax(res), max(res)
class App(tk.Tk):
    def __init__(self):
        tk.Tk.__init__(self)
        self.x = self.y = 0
        # Creating elements
        self.canvas = tk.Canvas(self, width=300, height=300, bg = "white", cursor="cross")
        self.label = tk.Label(self, text="Thinking..", font=("Helvetica", 48))
        self.classify_btn = tk.Button(self, text = "Recognise", command =         self.classify_handwriting) 
        self.button_clear = tk.Button(self, text = "Clear", command = self.clear_all)
        # Grid structure
        self.canvas.grid(row=0, column=0, pady=2, sticky=W, )
        self.label.grid(row=0, column=1,pady=2, padx=2)
        self.classify_btn.grid(row=1, column=1, pady=2, padx=2)
        self.button_clear.grid(row=1, column=0, pady=2)
        #self.canvas.bind("<Motion>", self.start_pos)
        self.canvas.bind("<B1-Motion>", self.draw_lines)
    def clear_all(self):
        self.canvas.delete("all")
    def classify_handwriting(self):
        HWND = self.canvas.winfo_id() # get the handle of the canvas
        rect = win32gui.GetWindowRect(HWND) # get the coordinate of the canvas
        im = ImageGrab.grab(rect)
        digit, acc = predict_digit(im)
        self.label.configure(text= str(digit)+', '+ str(int(acc*100))+'%')
    def draw_lines(self, event):
        self.x = event.x
        self.y = event.y
        r=8
        self.canvas.create_oval(self.x-r, self.y-r, self.x + r, self.y + r, fill='black')
app = App()
mainloop()

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 70ms/step


C:\Users\manan\anaconda3\Lib\site-packages\keras\src\models\functional.py:241: UserWarning: The structure of `inputs` doesn't match the expected structure.
Expected: input_layer
Received: inputs=('Tensor(shape=(1, 28, 28, 1))',)
  warnings.warn(msg)


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 32ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step


## Conclusion

In this project, I have successfully built a handwritten digit recognition system using a Convolutional Neural Network.

Key achievements:
- Achieved ~85% test accuracy
- Implemented CNN for image classification
- Built an interactive GUI for real-time predictions

This project demonstrates the power of deep learning in image recognition tasks.